google/cloud/security/scanner/audit/instance_network_interface_rules_engine.py

```diff
@@ -0,0 +1,222 @@
+                    'network': self,
+                    'is_external': boolean
+                }
+                whitelist {
+                    'network': ['project'...]
+
+                }
+            }
+
+        Args:
+            rule_def: A dictionary containing rule definition properties.
+            rule_index: The index of the rule from the rule definitions.
+                Assigned automatically when the rule book is built.
+
+        Args:
+            rule_def: A dictionary containing rule definition properties.
+            rule_index: The index of the rule from the rule definitions.
+            Assigned automatically when the rule book is built.
+        """
+        project = rule_def.get('project')
+        network = rule_def.get('network')
+        whitelist = rule_def.get('whitelist')
+        is_external_network = rule_def.get('is_external_network')
+
+        if (whitelist is None) or (project is None) or (network is None)\
+                or (is_external_network is None):
+            raise audit_errors.InvalidRulesSchemaError(
+                'Faulty rule {}'.format(rule_def.get('name')))
+
+        rule_def_resource = {'whitelist': whitelist,
+                             'project': escape_and_globify(project),
+                             'network': escape_and_globify(network),
+                             'is_external_network': is_external_network}
+
+        rule = Rule(rule_name=rule_def.get('name'),
+                    rule_index=rule_index,
+                    rules=rule_def_resource)
+
+        resource_rules = self.resource_rules_map.get(rule_index)
+        if not resource_rules:
+            self.resource_rules_map[rule_index] = rule
+
+    def get_resource_rules(self):
+        """Get all the resource rules."""
+        return self.resource_rules_map.values()
+
+
+class Rule(object):
+    """The rules class for instance_network_interface."""
+
+    def __init__(self, rule_name, rule_index, rules):
+        """Initialize.
+
+        Args:
+            rule_name: Name of the loaded rule
+            rule_index: The index of the rule from the  definitions
+            rules: The resources associated with the rules like the whitelist
+        """
+        self.rule_name = rule_name
+        self.rule_index = rule_index
+        self.rules = rules
+
+    def find_violations(self, instance_network_interface):
+        """Raise violation is the ip is not in the whitelist.
+
+        Args:
+            instance_network_interface: InstanceNetworkInterface obj
+        """
+        network_and_project = re.search(
+            'compute\/v1\/projects\/([^\/]*).*networks\/([^\/]*)',
+            instance_network_interface.network)
+        project = network_and_project.group(1)
+        network = network_and_project.group(2)
+        is_external_network = instance_network_interface.access_configs \
+            is not None
+        if not self.rules['whitelist'].get(project):
```

Comment:
This is me being ignorant here, but I am not sure I understand how, if the project is "falsy", it means the IP is not in the whitelist. (Presumably, because it implies that the IP is None?)